feat/configuration-contains-value

rikkarth opened the pull request:
I have been working on a new small feature which aims to validate if a value is present in a configuration through a first match validation.

This is useful when looking for specific values in multiple configurations and often can be leveraged to return the right Configuration if a value is present.

It can also be used to validate the integrity of a Configuration.

It is natively present in java.util.Properties (see [JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/util/Hashtable.html#containsValue-java.lang.Object-)).

I tried to mimic the structure already implemented for `containsKey` but added different implementation in order to retrieve the match for a value.

The first tests pass for all Configuration types, but I want to add more tests after the first appreciation of this PR.

I used recursion in some implementations for some Configuration types to retrieve the match for a value.

```java
    @Override
    protected boolean containsValueInternal(final String value) {
        return contains(getKeys(), value);
    }

    private boolean contains(Iterator<String> keys, final String value) {
        if (keys.hasNext()) {
            String nextKey = keys.next();
            Object valueFromKey = getProperty(nextKey);

            if (valueFromKey.equals(value)) {
                return true;
            }

            contains(keys, value);
        }

        return false;
    }
```

Tests currently fail here but only on GitHub, when I test them locally they don't fail. I would appreciate some feedback on this if possible.
![image](https://github.com/apache/commons-configuration/assets/76500344/98cb1896-2f75-43db-b22c-acf27614b16e)

```
Error:  Failures: 
Error:    TestServletConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletConfiguration ==> expected: <true> but was: <false>
Error:    TestServletContextConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletContextConfiguration ==> expected: <true> but was: <false>
Error:    TestServletFilterConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletFilterConfiguration ==> expected: <true> but was: <false>
```

I'm happy for any feedback.

Commit:
feat: contains value initial implementation

Commit:
test: tests adjusted

Commit:
chore: remove temp class

Commit:
chore: lint fix

Commit:
fix: MapConfiguration containsValue

test: moved test

Commit:
chore: minor cleanup, renaming

Commit:
test: improved err message

garydgregory commented:
@rikkarth 
Make sure you run `mvn` by itself to verify that all checks pass.

rikkarth commented:
This is the error when running `mvn`. Is this some build specification I've missed?

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.20.0:cmp (default) on project commons-configuration2: There is at least one incompatibility: org.apache.commons.configuration2.AbstractConfiguration.containsValueInternal(java.lang.String):METHOD_ABSTRACT_ADDED_TO_CLASS,org.apache.commons.configuration2.ImmutableConfiguration.containsValue(java.lang.String):METHOD_ADDED_TO_INTERFACE -> [Help 1]
```

rikkarth commented:
Ok, so I'm having an error that I think is build related:

Can you help correct this issue?

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.20.0:cmp (default) 
on project commons-configuration2: There is at least one incompatibility: 
org.apache.commons.configuration2.AbstractConfiguration.containsValueInternal(java.lang.String):METHOD_ABSTRACT_ADDED_TO_CLASS,
org.apache.commons.configuration2.ImmutableConfiguration.containsValue(java.lang.String):METHOD_ADDED_TO_INTERFACE
```

garydgregory commented:
Hello @rikkarth 
It looks like this PR breaks binary compatibility based on the japicmp error you quote and there are also failing unit tests.

Your profile states you are a "Problem-solving Prodigy", so... either less boasting or more problem solving ;-)

See also:
- https://garygregory.wordpress.com/2020/06/14/how-we-handle-binary-compatibility-at-apache-commons/ 
- https://wiki.openjdk.org/display/csr/Kinds+of+Compatibility
- https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html

HTH

rikkarth commented:
@garydgregory , thank you for the feedback and providing the documentation.

I must rethink my profile description.

Claudenw commented on src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:
```diff
@@ -478,6 +478,21 @@ protected Boolean performOperation() throws SQLException {
+            @Override
+            protected Boolean performOperation() throws SQLException {
+                try (ResultSet rs = openResultSet(String.format(SQL_GET_PROPERTY, table, valueColumn), true, value)) {
+                    return rs.next();
```
If I read the SQL_GET_PROPERTY correctly as well as the what you are passing in as arguments, you have used up all the prepared statement parameters and the `true` value in this line should be `false`.  

**WARNING** I may be mistaken.

rikkarth replied:
Turned it off but defined a test to check for return.

It asserts correctly.

Thank you.

```java
@Test
    public void testContainsValue() throws ConfigurationException {
        final DatabaseConfiguration config = helper.setUpConfig();
        config.addPropertyDirect("test", "test1");

        assertTrue(config.containsValue("test1"));
    }
```

Claudenw commented on src/main/java/org/apache/commons/configuration2/JNDIConfiguration.java:
```diff
@@ -147,6 +147,26 @@ protected boolean containsKeyInternal(String key) {
+            contains(keys, value);
+        }
+
+        return false;
```
See comments for similar method above.

Claudenw commented on src/main/java/org/apache/commons/configuration2/web/BaseWebConfiguration.java:
```diff
@@ -68,6 +69,26 @@ protected boolean containsKeyInternal(final String key) {
+        }
+
+        return false;
+    }
```
See comments for similar method above.  

Do all of these classes implement a common interface or class where a common method might be placed?

rikkarth replied:
They do, I reduced the duplication in my next commit.

Claudenw reviewed (changes requested):
Thanks for your contribution.  There are a couple of issues I think need to be fixed.  You might also wait for another review before you make these changes.

Commit:
fix: moved contains impl to AbstractConfiguration

rikkarth commented on src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java:
```diff
@@ -455,6 +455,26 @@ protected boolean containsKeyInternal(final String key) {
+            Object valueFromKey = getProperty(nextKey);
+
+            if (valueFromKey.equals(value)) {
+                return true;
```
100% agree. Will perform the change.

Thank you.

rikkarth commented on src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java:
```diff
@@ -455,6 +455,26 @@ protected boolean containsKeyInternal(final String key) {
+
+        return false;
+    }
+
```
That was my initial idea (use the while). The functional part of me decided against KISS.

I will revert to a while and no need for suplier since method (now updated in last commit) is only called from one place.

Commit:
change: simplified contains in AbstractConfiguration.java

Commit:
change: adjusted containsValueInternal in DatabaseConfiguration

test: containsValue() in TestDatabaseConfiguration

garydgregory commented:
> @garydgregory , thank you for the feedback and providing the documentation.
> 
> I must rethink my profile description.

It's hard to read people online but it appears you have good sense of humor 😀

Commit:
docs: add javadoc

rikkarth commented:
> > @garydgregory , thank you for the feedback and providing the documentation.
> > I must rethink my profile description.
> 
> It's hard to read people online but it appears you have good sense of humor 😀

A bit of sense of humour and humility is always nice to have.

I'm very grateful that you and @Claudenw have taken the time to look at my PR, thank you.

As for the compatibility issue, I don't think I'm breaking compatibility by adding a new method to the API, so that means I don't have to bump the version from 2.x to 3.x..

I would like to ask, how are you bumping the version since from past PRs I've read I don't see any change to the pom.

rikkarth commented:
Ok so I believe this is the reason why I broke compatibility.

```
However, adding overloaded methods has the potential to change method resolution and thus change the signatures of the method call sites in the resulting class file. Whether or not such a change is problematic with respect to source compatibility depends on what semantics are required and how the different overloaded methods operate on the same inputs, which interacts with behavioral equivalence notions. For example, consider adding to Lib an overloading of foo which takes an int:

// Behaviorally equivalent source compatible change
public final class Lib {
  public double foo(double d) {
    return d * 2.0;
  }

  // New overloading
  public double foo(int i) {
    return i * 2.0;
  }
}
```

garydgregory commented:
@rikkarth
You need to understand how to not break binary compatibility: Please see "13.5.6. Interface Method Declarations" in the JLS at the link I've already provided.

Commit:
fix: incompatibility issue

rikkarth commented:
> @rikkarth You need to understand how to not break binary compatibility: Please see "13.5.6. Interface Method Declarations" in the JLS at the link I've already provided.

Thank you so much for pointing me in the right direction. First time contributing to a project with this specification, but after reading all this it makes a lot of sense.

I would like to ask if you don't mind looking at my default implementation.

I think if this method is not implemented by the Class implementing ImmutableConfiguration, the caller should get an Exception with this message.

Or should it just return false?

```java
    /**
     * Returns true if this configuration contains one or more keys to this value. This operation is more expensive than
     * the {@link #containsKey containsKey} method.
     *
     * @param value value whose presence in this configuration is to be tested
     * @return {@code true} if this configuration maps one or more keys to the specified value
     * @throws NullPointerException if the value is {@code null}
     * @throws UnsupportedOperationException if this method is not supported by the implementing class.
     */
    default boolean containsValue(String value) {
        throw new UnsupportedOperationException(String.format("Class %s does not support this method.", this.getClass()));
    }
```

Commit:
docs: minor javadoc update

Commit:
test: nullity tests

Commit:
fix: add null validation to contains

test: nullity test adjusted

garydgregory commented:
@rikkarth 
If you think your PR is ready for review, then set it to "Ready for review" and I'll take a look over the next day or two.

Commit:
docs: javadoc contains in AbstractConfiguration

Commit:
docs: javadoc adaption

rikkarth commented:
Thank you for taking the time to look at my PR @garydgregory and @Claudenw .

I've submitted my PR for review as instructed.

Have a great day!

garydgregory commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -419,6 +419,20 @@ public final boolean containsKey(final String key) {
+     */
+    @Override
+    public final boolean containsValue(final String value) {
+        Objects.requireNonNull(value, "Parameter \"value\" cannot be null");
```
- Replace API calls of this pattern with simply:
`Objects.requireNonNull(value, "value");`
See the rest of the code base.
- Why does this check need to happen here at all? After all I don't get an NPE when I ask for a null value from a `HashMap`, for example ` new HashMap().containsValue(null)` returns false.

rikkarth replied:
containsValue delegates to `containsKeyInternal()`.

`containsKeyInternal()` implementation can be different depending on the implementing class.

If null is unavoidable we should always encourage safe null handling practices, such as using Objects.requireNonNull or conditional checks, to minimize the risk of (unexpected, unmanaged or uncontrolled) null pointer exceptions.

While `new HashMap().containsValue(null)` returns false, this implementation is inspired on `java.util.HashTable.contains()` and not HashMap which is also `java.util.Properties` implementation (extends HashTable).

```java
    // HashTable contains Implementation checks null at entry point
    public synchronized boolean contains(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        Entry<?,?> tab[] = table;
        for (int i = tab.length ; i-- > 0 ;) {
            for (Entry<?,?> e = tab[i] ; e != null ; e = e.next) {
                if (e.value.equals(value)) {
                    return true;
                }
            }
        }
        return false;
    }
```

I find this to be a much more coherent implementation since it doesn't fail silently and explicitely warns the user that the input value is null.

The reason for the "custom" message is so we can control exactly how containsValue fails.

What do you think?

rikkarth replied:
After discussion with @Claudenw I have removed the null validation. Please see comments.

Thank you :)

garydgregory commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -429,6 +443,22 @@ public final boolean containsKey(final String key) {
+     * @return {@code true} if and only if some key maps to the {@code value} argument in this hashtable as determined
+     * by the {@code equals} method; {@code false} otherwise.
+     * @throws NullPointerException if the value is {@code null}
+     */
```
New elements that are public or private should be documented with a Javadoc `@since` tag.

rikkarth replied:
How about protected ones? (sorry I know you didn't mention but just want to make sure)

Claudenw replied:
My approach is: if you are writing a javadoc add the `@since` annotation.

rikkarth replied:
Thank you I wasn't aware of this rule until this PR, thank you again for clarifying will do that from now on.

garydgregory commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -1517,6 +1547,25 @@ protected int sizeInternal() {
+     * @return true if the value is found in the properties, false otherwise
+     * @throws NullPointerException if keys or value is null
+     */
+    protected boolean contains(Iterator<String> keys, final String value) {
```
Use `final` where you can, here, `keys` can be `final`.

rikkarth replied:
Will do.

garydgregory commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -1517,6 +1547,25 @@ protected int sizeInternal() {
+     * @throws NullPointerException if keys or value is null
+     */
+    protected boolean contains(Iterator<String> keys, final String value) {
+        Objects.requireNonNull(keys, "Parameter \"keys\" cannot be null");
```
See my previous comment RE `Objects.requireNonNull()`.
Same comment RE null checking in the first place. Why not return false?

Claudenw replied:
See my previous comment re `Objects.requireNonNull()`

garydgregory commented on src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:
```diff
@@ -478,6 +478,21 @@ protected Boolean performOperation() throws SQLException {
+                }
+            }
+        };
+
```
No need for extra whitespace (blank lines). If you want to call out a code fragment, then you can use an inline `//` comment to alert the reader to something specific.

rikkarth replied:
Corrected.

garydgregory commented on src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:
```diff
@@ -62,6 +62,20 @@ public interface ImmutableConfiguration {
     boolean containsKey(String key);
 
+    /**
+     * Returns true if this configuration contains one or more matches to this value. This operation stops at first
```
Be consistent in your Javadoc comments. In this case, "contains" methods that return a boolean are "testing" for the presence of something, so you can start the comment with "Tests" like "Tests whether this or that happens".

rikkarth replied:
Thank you for the feedback. I have adjusted this.

garydgregory commented on src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:
```diff
@@ -62,6 +62,20 @@ public interface ImmutableConfiguration {
+     * @return {@code true} if this configuration maps one or more keys to the specified value
+     * @throws NullPointerException          if the value is {@code null}
+     * @throws UnsupportedOperationException if this method is not supported by the implementing class.
+     */
```
Missing Javadoc since tag.

rikkarth replied:
Done.

garydgregory commented on src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:
```diff
@@ -62,6 +62,20 @@ public interface ImmutableConfiguration {
+     * @throws NullPointerException          if the value is {@code null}
+     * @throws UnsupportedOperationException if this method is not supported by the implementing class.
+     */
+    default boolean containsValue(String value) {
```
Can this method be implemented in terms of other interface methods instead of throwing an exception?

rikkarth replied:
Ok, so I gave an actual default implementation to `containsValue`  in `ImmutableConfiguration` and tested it with that DummyClass that I didn't want to create to test it and see if it worked.

And it worked.

![image](https://github.com/apache/commons-configuration/assets/76500344/4604c968-fac6-4e6e-9779-a8e3524b54db)

![image](https://github.com/apache/commons-configuration/assets/76500344/869a88d0-ce77-4e7f-96c4-561cb50bd681)

![image](https://github.com/apache/commons-configuration/assets/76500344/206f1ca3-9a6b-403f-941f-8b357eb8c8bd)

garydgregory commented on src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:
```diff
@@ -62,6 +62,20 @@ public interface ImmutableConfiguration {
+     */
+    default boolean containsValue(String value) {
+        throw new UnsupportedOperationException(
+            String.format("Class %s does not support this method.", this.getClass()));
```
The message is redundant and bad, it read like "Class class com.example.Foo does not...".
You must have not tested this case. A simpler message could just say `getClass().getName() + ".myMethod"`

rikkarth replied:
Hey Greg, you are right

I did in fact not test this method since I couldn't think of a way to test this in a class that doesn't implement the method without creating a dummy class for it.

Do you think I should just adapt the message like you suggested?

Thank you.

garydgregory replied:
Yes.

garydgregory commented on src/test/java/org/apache/commons/configuration2/NonCloneableConfiguration.java:
```diff
@@ -45,6 +45,11 @@ protected boolean containsKeyInternal(final String key) {
         return false;
     }
 
+    @Override
```
Javadoc

rikkarth replied:
Done.

garydgregory commented on src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:
```diff
@@ -78,6 +78,49 @@ public void testAddPropertyDirect() {
 
+    @Test
+    void testContainsValue(){
+        Configuration config = getConfiguration();
```
Use `final`.

rikkarth replied:
Done.

garydgregory commented on src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:
```diff
@@ -78,6 +78,49 @@ public void testAddPropertyDirect() {
+    @Test
+    void testContains(){
+        AbstractConfiguration config = getConfiguration();
+
```
No need to blank lines. See my previous comment.

rikkarth replied:
Corrected.

garydgregory commented on src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:
```diff
@@ -78,6 +78,49 @@ public void testAddPropertyDirect() {
     }
 
+    @Test
+    void testContainsValue(){
```
The style of the component is to use `public` test methods prefixed with "test" in the name and camel-case. Look at other tests for examples.

rikkarth replied:
Will correct.

garydgregory reviewed (changes requested):
@rikkarth 
Thank you for your PR.
Please see my comments.

Commit:
feat: add impl to containsValue in ImmutableConfiguration

chore: add final to contains in AbstractConfiguration
chore: remove white space in DatabaseConfiguration
chore: add missing public and final to TestAbstractConfiguration
test: given existing value test containsValue with default impl

Commit:
docs: javadoc NonCloneableConfiguration

Claudenw commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -419,6 +419,20 @@ public final boolean containsKey(final String key) {
+     */
+    @Override
+    public final boolean containsValue(final String value) {
+        Objects.requireNonNull(value, "Parameter \"value\" cannot be null");
```
I think the `Objects.requiredNull` should be removed since this is an AbstractConfiguration and my actually be implementing a configuration that allows nulls.  It may make sense in final classes to check for null but not here.

rikkarth replied:
I agree, that makes sense, since `containsValue` delegates to containsValueInternal there could be an implementation that allows null values.

Will remove.

rikkarth replied:
If every configuration can have a different implementation for containsValue it means that every configuration must be tested for their (existing) implementation of containsValue.

Is it ok if I do it all in this PR?

Claudenw reviewed (changes requested):
Thanks for your work on this.  It is looking better.

Commit:
docs: complete javadocs

Commit:
test: adjusted containsValue tests, test each containsValue implementation

garydgregory reviewed (changes requested):
Hello @rikkarth 
- Please rebase on git master to pick up the latest Checkstyle configuration and fixes.
- See my comments in this PR
- Run `mvn` by itself to run all build checks

TY!

rikkarth commented on src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:
```diff
@@ -419,6 +419,20 @@ public final boolean containsKey(final String key) {
 
+    /**
+     * {@inheritDoc} This implementation handles synchronization and delegates to {@code containsKeyInternal()}.
+     * @since 2.0
```
Do you mean the version value? Or the tag?

garydgregory replied:
New public and protected method should have a Javadoc since tag. Semantic versioning makes the next version 2.11.0.

rikkarth commented on src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:
```diff
@@ -78,6 +78,35 @@ public void testAddPropertyDirect() {
+    @Test
+    public void testContainsValue(){
+        final Configuration config = getConfiguration();
+        assertTrue(config.containsValue("value1"), "should return true for class " + this.getClass());
```
Will be removed.

Commit:
test: remove string messages from tests

Commit:
chore: changed tests signature names

Commit:
docs: adjusted javadoc

Commit:
docs: adjusted since javadoc version

Commit:
docs: add missing javadoc

rikkarth commented:
> Hello @rikkarth
> 
> * Please rebase on git master to pick up the latest Checkstyle configuration and fixes.
> * See my comments in this PR
> * Run `mvn` by itself to run all build checks
> 
> TY!

Done as requested.

While changing all the since tags and adjusting tests signature names I was wondering how dull must have been for you to review this, sorry about that. In future PRs I will know better the design, patterns and style applied to this component from the experience acquired in this first PR.

Thank you for your time again @garydgregory

garydgregory commented on .github/workflows/coverage.yml:
```diff
@@ -42,6 +42,6 @@ jobs:
 
     - name: Upload coverage to Codecov
-      uses: codecov/codecov-action@84508663e988701840491b86de86b666e8a86bed # v3
+      uses: codecov/codecov-action@5ecb98a3c6b747ed38dc09f787459979aebb39be # v3
```
The PR should not contain this change.

garydgregory commented on .github/workflows/maven.yml:
```diff
@@ -27,10 +27,10 @@ jobs:
     strategy:
       matrix:
-        java: [ 8, 11, 17, 21 ]
+        java: [ 8, 11, 17, 21, 22 ]
```
The PR should not contain this change.

garydgregory reviewed (changes requested):
The PR should not contain a bunch of these changed files. Please rollback those changes to only contain the required changes. ty!

rikkarth commented:
Done as instructed.

garydgregory reviewed (changes requested):
@rikkarth 
Something is still messed up in this PR. It touches 67 files! It looks like it contains unrelated changes from master. Something must have gone wrong in whatever git commands you used. The PR should only contain _your_ changes.
TY!

rikkarth commented:
I'll fix it. Had to change computers and my gpg was not signing commits.

rikkarth commented:
I believe it's done now @garydgregory :) had to do some git gymnastics «

Commit:
workflows update

garydgregory commented on .github/workflows/coverage.yml:
```diff
@@ -42,6 +42,6 @@ jobs:
 
     - name: Upload coverage to Codecov
-      uses: codecov/codecov-action@5ecb98a3c6b747ed38dc09f787459979aebb39be # v3
+      uses: codecov/codecov-action@84508663e988701840491b86de86b666e8a86bed # v3
```
This change does not belong in this PR.

garydgregory commented on .github/workflows/maven.yml:
```diff
@@ -27,7 +27,7 @@ jobs:
     strategy:
       matrix:
-        java: [ 8, 11, 17, 21, 22 ]
+        java: [ 8, 11, 17, 21 ]
```
This change does not belong in this PR.

garydgregory reviewed (changes requested):
Two changes do not belong in this PR. See my comments. TY!

Commit:
Revert "workflows update"

This reverts commit 56fafc9e972aa901c34c3157d9ea72537db8b1b4.

rikkarth commented:
I was under the wrong impression you wanted those changes, but I reverted them now!

garydgregory committed:
Add Javadoc since tag
